### ISSUE TITLE
Pylance server keep on restarting on big project

### ISSUE THREAD

**Issue by Sommary** [microsoft/pylance-release#3408]
## Environment data

-   Language Server version: v2022.9.41
-   OS and version: Ubuntu 20.04.5 LTS, 64-bit, Memory 31,2 GiB, Intel® Core™ i5-8500 CPU @ 3.00GHz × 6
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10 and several projects on 3.6-3.8

## Repro Steps

1. Launch VS Code, 
2. Open a workspace with a lot of microservices (I have 105 folders and each is a project itself) 
3. Wait for Python Language Server to start.

## Expected behavior

Server starts.

## Actual behavior

Server either keep on restarting or report it will not due to reoccurring error. The advice from https://github.com/microsoft/pylance-release/issues/3307 to use "openFilesOnly" did not help. If change "python.analysis.logLevel" to "Trace" while server keeps restarting, it will crush and will report "The server will not be restarted". If relaunch VS Code, it automatically opens previously opened project (my 105 folder workspace) and keeps trying to restart the language server forever.
Unfortunately, I cannot provide you with the full log, as the project is proprietary. 

## Logs

```
<--- Last few GCs --->

[386654:0x140c0274c000]    46915 ms: Mark-sweep (reduce) 10.4 (15.1) -> 10.4 (12.1) MB, 194.4 / 0.0 ms  (average mu = 0.063, current mu = 0.001) last resort; GC in old space requested
[386654:0x140c0274c000]    47108 ms: Mark-sweep (reduce) 10.4 (12.1) -> 10.3 (12.1) MB, 193.0 / 0.0 ms  (average mu = 0.024, current mu = 0.000) last resort; GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
[Info  - 11:09:04 AM] Connection to server got closed. Server will restart.
[Info  - 11:09:05 AM] (389994) Pylance language server 2022.9.41 (pyright 59bdc42e) starting
```

```

<--- Last few GCs --->

[420468:0x25f4025e4000]    42784 ms: Mark-sweep (reduce) 11.1 (15.9) -> 11.1 (12.9) MB, 99.1 / 0.0 ms  (average mu = 0.024, current mu = 0.003) last resort; GC in old space requested
[420468:0x25f4025e4000]    42804 ms: Mark-sweep (reduce) 11.1 (12.9) -> 11.1 (12.9) MB, 19.5 / 0.0 ms  (average mu = 0.019, current mu = 0.001) last resort; GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
[Error - 11:16:47 AM] The Pylance server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
```

**Comment by erictraut** [comment]
Please create a `pyrightconfig.json` file in your project's root directory and add the following to this file:

```json
{
    "verboseOutput": true
}
```

Then reload your project and reproduce the out-of-memory condition. The output should now include more details. Look for lines in the output logs that start with "Heap stats: ". Please provide the last few of these heap stat outputs prior to the crash.

**Comment by Sommary** [comment]
Adding `pyrightconfig.json` with suggested setting did not help much as there are no lines start with "Heap stats: ". I tried to place this config into directory which is being opened as multi-root workspace (`/home/sommary/work/root/<my_company_name>`) and in its parent directory (`/home/sommary/work/root`) also.

So I've collected full logs. In them I have replaced my company name and projects names. I hope this will let you figure out the reason of the issue.
[python_language_server_output_redacted.txt](https://github.com/microsoft/pylance-release/files/9683049/python_language_server_output_redacted.txt)

**Comment by erictraut** [comment]
OK, the fact that you're using a multi-root workspace is a really important detail that you didn't previously mention. How many workspace roots do you have in your project?

When pylance is used in a multi-root workspace configuration, it creates a new instances of the type checker service for each root. Each service instance manages its own in-memory cache of types as it performs type checking, and it empties this cache if it determines that its cache could overflow the available heap space. However, each type checker service instance doesn't have knowledge of the other instances. I suspect what's happening here is that the sum total of memory used by all of the type checker service instances is consuming all of the memory in the heap without any one service instance knowing that it is in danger of doing so.

We'll need to discuss potential mitigations or workarounds. In the meantime, potential workarounds might include reducing the number of workspace roots (and potentially even switching to a single-root workspace), adding a pyrightconfig.json in a subset of the root directories that includes an `"exclude": "."` — effectively telling the type checker not to attempt any type checking for any files under that root, or opening a single workspace root at a time in your editor.

**Comment by Sommary** [comment]
There are like 100 workspace roots in the project. Using multi-root workspace is more appropriate than opening a single workspace root because I often work with several projects at the same time in a day. And the next day my task may involve other combination of projects.
I'll try to follow you advice about excluding some roots and let you know.
Do you consider to rework current logic of type checker for multi-root workspace in not so distant future?

**Comment by erictraut** [comment]
I've added some additional logic to track heap consumption across multiple type evaluator service instances. This should prevent the crashes you're seeing when using pylance with a project with many workspace roots.

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2022.10.11, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20221011-5-October-2022-PreRelease)
